File: backend-challenge/src/loaders/couponLoader.ts
import fs from 'fs';
import path from 'path';
import os from 'os';
import zlib from 'zlib';
import readline from 'readline';
import { config } from '../config';
import { logger } from '../utils/logger';

const NUM_BUCKETS = 256;     // 256 bucket files per gz file → 768 temp files total
const MIN_LEN = 8;
const MAX_LEN = 10;
const FLUSH_THRESHOLD = 10_000; // lines buffered per bucket before flushing to disk

/**
 * Hash a code to a bucket index in [0, NUM_BUCKETS).
 */
function hashBucket(code: string): number {
  let h = 0;
  for (let i = 0; i < code.length; i++) {
    h = (Math.imul(31, h) + code.charCodeAt(i)) | 0;
  }
  return (h >>> 0) % NUM_BUCKETS;
}

function popcount(n: number): number {
  let c = 0;
  while (n) { c += n & 1; n >>>= 1; }
  return c;
}

/**
 * Two-phase coupon loader that handles files with hundreds of millions of unique codes.
 *
 * Phase 1 — distribute:
 *   Stream each gz file once, filter for 8-10 char codes, and append each code
 *   to a per-bucket temp file (f{fileIdx}_b{bucket}.txt).
 *
 * Phase 2 — reduce:
 *   For each bucket, load the 3 small bucket files into a bitmask Map,
 *   collect codes with popcount(mask) >= 2, then discard the Map.
 *   At most NUM_BUCKETS unique codes are in memory at once.
 */
export async function loadCoupons(): Promise<Set<string>> {
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'kart-coupons-'));
  logger.info(`Coupon temp dir: ${tmpDir}`);

  try {
    // ── Phase 1: distribute ──────────────────────────────────────────────────
    for (let fi = 0; fi < config.couponFiles.length; fi++) {
      const filePath = path.join(config.couponDir, config.couponFiles[fi]);

      if (!fs.existsSync(filePath)) {
        logger.warn(`Coupon file not found, skipping: ${filePath}`);
        continue;
      }

      logger.info(`Phase 1 [${fi + 1}/${config.couponFiles.length}]: distributing ${config.couponFiles[fi]}`);
      await distributeFile(filePath, fi, tmpDir);
    }

    // ── Phase 2: reduce ──────────────────────────────────────────────────────
    logger.info(`Phase 2: reducing ${NUM_BUCKETS} buckets...`);
    const validCoupons = new Set<string>();

    for (let b = 0; b < NUM_BUCKETS; b++) {
      const bitmask = new Map<string, number>();

      for (let fi = 0; fi < config.couponFiles.length; fi++) {
        const bucketFile = path.join(tmpDir, `f${fi}_b${b}.txt`);
        if (!fs.existsSync(bucketFile)) continue;

        const bit = 1 << fi;
        const content = fs.readFileSync(bucketFile, 'utf8');
        for (const code of content.split('\n')) {
          if (!code) continue;
          bitmask.set(code, (bitmask.get(code) ?? 0) | bit);
        }
      }

      for (const [code, mask] of bitmask) {
        if (popcount(mask) >= 2) validCoupons.add(code);
      }

      if ((b + 1) % 64 === 0) {
        logger.info(
          `Phase 2: ${b + 1}/${NUM_BUCKETS} buckets done, valid so far: ${validCoupons.size.toLocaleString()}`,
        );
      }
    }

    logger.info(`Coupon loading complete: ${validCoupons.size.toLocaleString()} valid coupons`);
    return validCoupons;
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
    logger.info(`Cleaned up temp dir: ${tmpDir}`);
  }
}

/**
 * Stream a single gz file, filter 8-10 char codes, and write each code to the
 * appropriate bucket temp file. In-memory line buffers are flushed to disk when
 * they hit FLUSH_THRESHOLD to avoid exhausting RAM.
 */
async function distributeFile(filePath: string, fileIdx: number, tmpDir: string): Promise<void> {
  const buffers: string[][] = Array.from({ length: NUM_BUCKETS }, () => []);

  const flushBucket = (b: number): void => {
    if (buffers[b].length === 0) return;
    const dest = path.join(tmpDir, `f${fileIdx}_b${b}.txt`);
    fs.appendFileSync(dest, buffers[b].join('\n') + '\n');
    buffers[b] = [];
  };

  return new Promise<void>((resolve, reject) => {
    let lineCount = 0;

    const fileStream = fs.createReadStream(filePath);
    const gunzip = zlib.createGunzip();
    const rl = readline.createInterface({ input: gunzip, crlfDelay: Infinity });

    fileStream.on('error', reject);
    gunzip.on('error', reject);

    rl.on('line', (line) => {
      const code = line.trim();
      // Only codes that can ever be valid need to be stored
      if (code.length < MIN_LEN || code.length > MAX_LEN) return;

      const b = hashBucket(code);
      buffers[b].push(code);
      if (buffers[b].length >= FLUSH_THRESHOLD) flushBucket(b);

      lineCount++;
      if (lineCount % 5_000_000 === 0) {
        logger.info(`  File ${fileIdx + 1}: distributed ${lineCount.toLocaleString()} valid-length lines`);
      }
    });

    rl.on('close', () => {
      for (let b = 0; b < NUM_BUCKETS; b++) flushBucket(b);
      logger.info(`  File ${fileIdx + 1}: ${lineCount.toLocaleString()} valid-length lines distributed`);
      resolve();
    });

    rl.on('error', reject);
    fileStream.pipe(gunzip);
  });
}

File: go-backend/internal/loader/coupon_loader.go
// Package loader implements the two-phase hash-partitioned coupon loading
// strategy that keeps memory usage flat even for multi-GB gz input files.
package loader

import (
	"bufio"
	"compress/gzip"
	"fmt"
	"hash/fnv"
	"math/bits"
	"os"
	"path/filepath"
	"runtime"
	"sync"

	"go.uber.org/zap"
)

const (
	numBuckets   = 256
	minCodeLen   = 8
	maxCodeLen   = 10
	minPopcount  = 2 // a code must appear in at least 2 files to be considered valid
)

// LoadCoupons performs the two-phase loading and returns a set of valid codes.
//
// Phase 1: stream each *.gz file, filter lines whose length is in [8,10],
// write each qualifying code to one of 256 bucket temp files based on
// fnv32a(code) % 256.
//
// Phase 2: process every bucket file in parallel (goroutine pool = NumCPU),
// read all codes, build a bitmask per unique code (bit N = seen in file N),
// collect codes whose popcount >= minPopcount.
func LoadCoupons(dir string, logger *zap.Logger) (map[string]struct{}, error) {
	// Locate gz files.
	pattern := filepath.Join(dir, "*.gz")
	gzFiles, err := filepath.Glob(pattern)
	if err != nil {
		return nil, fmt.Errorf("glob %q: %w", pattern, err)
	}
	if len(gzFiles) == 0 {
		logger.Warn("no .gz coupon files found; starting with empty coupon set", zap.String("dir", dir))
		return make(map[string]struct{}), nil
	}
	if len(gzFiles) > 64 {
		return nil, fmt.Errorf("too many gz files (%d); max 64 supported for bitmask", len(gzFiles))
	}

	logger.Info("phase 1: partitioning coupon files into buckets",
		zap.Int("files", len(gzFiles)),
		zap.Int("buckets", numBuckets),
	)

	// Create temp directory for bucket files.
	tmpDir, err := os.MkdirTemp("", "coupon-buckets-*")
	if err != nil {
		return nil, fmt.Errorf("create temp dir: %w", err)
	}
	defer func() {
		if removeErr := os.RemoveAll(tmpDir); removeErr != nil {
			logger.Warn("failed to remove temp dir", zap.String("dir", tmpDir), zap.Error(removeErr))
		}
	}()

	// Phase 1 — stream each file, fan-out into buckets.
	for fileIdx, gzPath := range gzFiles {
		if err := phase1File(gzPath, fileIdx, tmpDir); err != nil {
			return nil, fmt.Errorf("phase1 %q: %w", gzPath, err)
		}
		logger.Info("phase 1 file done", zap.String("file", filepath.Base(gzPath)))
	}

	logger.Info("phase 2: building valid set from buckets",
		zap.Int("parallelism", runtime.NumCPU()),
	)

	// Phase 2 — parallel bucket processing.
	valid := make(map[string]struct{})
	var mu sync.Mutex

	concurrency := runtime.NumCPU()
	sem := make(chan struct{}, concurrency)
	var wg sync.WaitGroup
	var firstErr error
	var errMu sync.Mutex

	for b := 0; b < numBuckets; b++ {
		bucketPath := filepath.Join(tmpDir, fmt.Sprintf("bucket-%03d.txt", b))

		// Skip buckets that were never written.
		if _, statErr := os.Stat(bucketPath); os.IsNotExist(statErr) {
			continue
		}

		wg.Add(1)
		sem <- struct{}{}
		go func(path string, fileCount int) {
			defer wg.Done()
			defer func() { <-sem }()

			codes, processErr := phase2Bucket(path, fileCount)
			if processErr != nil {
				errMu.Lock()
				if firstErr == nil {
					firstErr = processErr
				}
				errMu.Unlock()
				return
			}

			if len(codes) > 0 {
				mu.Lock()
				for c := range codes {
					valid[c] = struct{}{}
				}
				mu.Unlock()
			}
		}(bucketPath, len(gzFiles))
	}

	wg.Wait()

	if firstErr != nil {
		return nil, fmt.Errorf("phase2: %w", firstErr)
	}

	logger.Info("coupon loading complete", zap.Int("valid_codes", len(valid)))
	return valid, nil
}

// phase1File streams a single gz file and writes each qualifying code to the
// appropriate bucket file.  bucket files are opened in append mode so that
// multiple source files accumulate into the same set of buckets.
func phase1File(gzPath string, fileIdx int, tmpDir string) error {
	f, err := os.Open(gzPath)
	if err != nil {
		return err
	}
	defer f.Close()

	gr, err := gzip.NewReader(f)
	if err != nil {
		return fmt.Errorf("gzip reader: %w", err)
	}
	defer gr.Close()

	// Open / cache bucket writers for this file so we don't thrash the OS.
	writers := make([]*bufio.Writer, numBuckets)
	files := make([]*os.File, numBuckets)
	defer func() {
		for i, bw := range writers {
			if bw != nil {
				_ = bw.Flush()
			}
			if files[i] != nil {
				_ = files[i].Close()
			}
		}
	}()

	getBucket := func(idx int) (*bufio.Writer, error) {
		if writers[idx] != nil {
			return writers[idx], nil
		}
		p := filepath.Join(tmpDir, fmt.Sprintf("bucket-%03d.txt", idx))
		bf, openErr := os.OpenFile(p, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o600)
		if openErr != nil {
			return nil, openErr
		}
		files[idx] = bf
		writers[idx] = bufio.NewWriterSize(bf, 1<<20) // 1 MiB write buffer
		return writers[idx], nil
	}

	h := fnv.New32a()
	scanner := bufio.NewScanner(gr)
	scanner.Buffer(make([]byte, 1<<20), 1<<20)

	for scanner.Scan() {
		code := scanner.Text()
		l := len(code)
		if l < minCodeLen || l > maxCodeLen {
			continue
		}

		h.Reset()
		_, _ = h.Write([]byte(code))
		bucketIdx := int(h.Sum32() % numBuckets)

		bw, bucketErr := getBucket(bucketIdx)
		if bucketErr != nil {
			return bucketErr
		}
		// Format: "<fileIdx>:<code>\n"
		if _, writeErr := fmt.Fprintf(bw, "%d:%s\n", fileIdx, code); writeErr != nil {
			return writeErr
		}
	}

	return scanner.Err()
}

// phase2Bucket reads a bucket file, builds per-code bitmasks, and returns
// the set of codes that appear in at least minPopcount distinct source files.
func phase2Bucket(bucketPath string, fileCount int) (map[string]struct{}, error) {
	f, err := os.Open(bucketPath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	// bitmask uses uint64; supports up to 64 source files.
	bitmasks := make(map[string]uint64)

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1<<16), 1<<16)

	for scanner.Scan() {
		line := scanner.Text()
		// Parse "<fileIdx>:<code>"
		colonIdx := -1
		for i := 0; i < len(line); i++ {
			if line[i] == ':' {
				colonIdx = i
				break
			}
		}
		if colonIdx < 0 {
			continue
		}
		fileIdxStr := line[:colonIdx]
		code := line[colonIdx+1:]

		// Parse file index manually (avoid strconv import overhead in hot path).
		fileIdx := 0
		for _, ch := range fileIdxStr {
			if ch < '0' || ch > '9' {
				fileIdx = -1
				break
			}
			fileIdx = fileIdx*10 + int(ch-'0')
		}
		if fileIdx < 0 || fileIdx >= fileCount {
			continue
		}

		bitmasks[code] |= 1 << uint(fileIdx)
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	result := make(map[string]struct{})
	for code, mask := range bitmasks {
		if bits.OnesCount64(mask) >= minPopcount {
			result[code] = struct{}{}
		}
	}
	return result, nil
}
